src/Type/Doctrine/Query/QueryResultDynamicReturnTypeExtension.php

```diff
@@ -70,14 +72,27 @@ public function getTypeFromMethodCall(
 
 		$queryType = $scope->getType($methodCall->var);
 		$queryResultType = $this->getQueryResultType($queryType);
+		$queryIndexType = $this->getQueryIndexType($queryType);
 
 		return $this->getMethodReturnTypeForHydrationMode(
 			$methodReflection,
 			$hydrationMode,
-			$queryResultType
+			$queryResultType,
+			$queryIndexType
 		);
 	}
 
+	private function getQueryIndexType(Type $queryType): Type
+	{
+		if (!$queryType instanceof GenericObjectType) {
```

Comment:
I think this should rather be `PHPStan\Type\Doctrine\Query\QueryType`, but this was used below.